src/parquet/column/test-util.h

```diff
@@ -177,39 +246,171 @@ void DataPageBuilder<Type::BOOLEAN>::AppendValues(const ColumnDescriptor *d,
-    const std::vector<int16_t>& def_levels, int16_t max_def_level,
-    const std::vector<int16_t>& rep_levels, int16_t max_rep_level) {
+template <typename Type>
+static shared_ptr<DataPage> MakeDataPage(const ColumnDescriptor *d,
+    const vector<typename Type::c_type>& values, int num_vals,
+    Encoding::type encoding, const uint8_t* indices, int indices_size,
+    const vector<int16_t>& def_levels, int16_t max_def_level,
+    const vector<int16_t>& rep_levels, int16_t max_rep_level) {
+  int num_values = 0;
+
   InMemoryOutputStream page_stream;
-  test::DataPageBuilder<TYPE> page_builder(&page_stream);
+  test::DataPageBuilder<Type> page_builder(&page_stream);
 
   if (!rep_levels.empty()) {
     page_builder.AppendRepLevels(rep_levels, max_rep_level);
   }
-
   if (!def_levels.empty()) {
     page_builder.AppendDefLevels(def_levels, max_def_level);
   }
 
-  page_builder.AppendValues(d, values);
+  if (encoding == Encoding::PLAIN) {
+    page_builder.AppendValues(d, values, encoding);
+    num_values = page_builder.num_values();
+  } else {//DICTIONARY PAGES
+    page_stream.Write(indices, indices_size);
+    num_values = std::max(page_builder.num_values(), num_vals);
+  }
 
   auto buffer = page_stream.GetBuffer();
 
-  return std::make_shared<DataPage>(buffer, page_builder.num_values(),
-      page_builder.encoding(),
+  return std::make_shared<DataPage>(buffer, num_values,
+      encoding,
       page_builder.def_level_encoding(),
       page_builder.rep_level_encoding());
 }
 
-template <int TYPE, typename T>
-static void Paginate(const ColumnDescriptor *d,
-    const std::vector<T>& values,
-    const std::vector<int16_t>& def_levels, int16_t max_def_level,
-    const std::vector<int16_t>& rep_levels, int16_t max_rep_level,
-    int num_levels_per_page, const std::vector<int>& values_per_page,
-    std::vector<std::shared_ptr<Page> >& pages) {
+
+template <typename TYPE>
+class DictionaryPageBuilder {
+ public:
+  typedef typename TYPE::c_type TC;
+  static constexpr int TN = TYPE::type_num;
+
+  // This class writes data and metadata to the passed inputs
+  explicit DictionaryPageBuilder(const ColumnDescriptor *d,
+      InMemoryOutputStream* sink, Encoding::type encoding = Encoding::RLE_DICTIONARY) :
+      sink_(sink),
+      num_dict_values_(0),
+      encoding_(Encoding::RLE_DICTIONARY),
+      have_values_(false) {
+        int type_length = 0;
+        if (TN == Type::FIXED_LEN_BYTE_ARRAY) {
+          type_length = d->type_length();
+        }
+        dict_buffer_ = std::make_shared<OwnedMutableBuffer>();
+        encoder_.reset(new DictEncoder<TC>(&pool_, type_length));
+        encoding_ = encoding;
+  }
+
+  void AppendValues(const vector<TC>& values) {
+    int num_values = values.size();
+    // Dictionary encoding
+    for (int i = 0; i < num_values; ++i) {
+      encoder_->Put(values[i]);
+    }
+    dict_buffer_->Resize(encoder_->dict_encoded_size());
+    encoder_->WriteDict(dict_buffer_->mutable_data());
```

Comment:
If we want a batch of dictionary data, we have to change this API to take an offset.